diff --git a/hw/pci/pci.c b/hw/pci/pci.c
index 24fae16..eeabebf 100644
--- a/hw/pci/pci.c
+++ b/hw/pci/pci.c
@@ -60,6 +60,8 @@ static Property pci_props[] = {
     DEFINE_PROP_UINT32("rombar",  PCIDevice, rom_bar, 1),
     DEFINE_PROP_BIT("multifunction", PCIDevice, cap_present,
                     QEMU_PCI_CAP_MULTIFUNCTION_BITNR, false),
+    DEFINE_PROP_UINT32("subvendor_id", PCIDevice, subvendor_id, 0),
+    DEFINE_PROP_UINT32("subsystem_id", PCIDevice, subsystem_id, 0),
     DEFINE_PROP_BIT("command_serr_enable", PCIDevice, cap_present,
                     QEMU_PCI_CAP_SERR_BITNR, true),
     DEFINE_PROP_BIT("x-pcie-lnksta-dllla", PCIDevice, cap_present,
@@ -1005,7 +1007,12 @@ static PCIDevice *do_pci_register_device(PCIDevice *pci_dev, PCIBus *bus,
     pci_config_set_class(pci_dev->config, pc->class_id);
 
     if (!pc->is_bridge) {
-        if (pc->subsystem_vendor_id || pc->subsystem_id) {
+        if (pci_dev->subvendor_id || pci_dev->subsystem_id) {
+            pci_set_word(pci_dev->config + PCI_SUBSYSTEM_VENDOR_ID,
+                         pci_dev->subvendor_id);
+            pci_set_word(pci_dev->config + PCI_SUBSYSTEM_ID,
+                         pci_dev->subsystem_id);
+        } else if (pc->subsystem_vendor_id || pc->subsystem_id) {
             pci_set_word(pci_dev->config + PCI_SUBSYSTEM_VENDOR_ID,
                          pc->subsystem_vendor_id);
             pci_set_word(pci_dev->config + PCI_SUBSYSTEM_ID,
diff --git a/include/hw/pci/pci.h b/include/hw/pci/pci.h
index 772692f..9233bac 100644
--- a/include/hw/pci/pci.h
+++ b/include/hw/pci/pci.h
@@ -341,6 +341,9 @@ struct PCIDevice {
     MSIVectorUseNotifier msix_vector_use_notifier;
     MSIVectorReleaseNotifier msix_vector_release_notifier;
     MSIVectorPollNotifier msix_vector_poll_notifier;
+
+    uint32_t subvendor_id;
+    uint32_t subsystem_id;
 };
 
 void pci_register_bar(PCIDevice *pci_dev, int region_num,
